src/Core/Domain/Country/ValueObject/NoCountryId.php

```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Country\ValueObject;
+
+/**
+ * Provides country id value
+ */
+class NoCountryId implements CountryIdInterface
```

Comment:
I don't like the smell of this class which does nothing.
Let's end the review to conclude.

Reply:
HI @MeKeyCool 

all these modifications are on purpose we introduce this concept of `NoEntityId` for a good reason. We need to be able to identify clearly that a relationship must be set to "no relationship" For example in the specific price command you can set some "filters" based on a few criteria.

As such we need a value to identify the no relation, we can't rely on null because most of the setter in product commands use null to indicate that there is no updates. So we need those three use case:
- null: no modifications
- EntityId: set the relationship with specified entity
- NoEntityId: remove the relationship to any entity

This could be handled via only an EntityId which accepts the value 0, and at first that's what we did. But it means the VO doesn't fulfill its whole purpose which is to ensure that the VO represents a valid identification of an entity, this it should always be > 0

Allowing the 0 value means the VO doesn't ensure the relationship, and quite often we need to handle some if cases:
```
if ($voId->getValue() === 0) {
} else {
}
```

By forcing the EntityId to never be 0 we ensure that we never need to check the 0 value, which simplifies the code in many places. It also semantically tells us something right away:
- EntityId represents a relationship to an existing entity, and it's required
- NoEntityId represents the lack of said relationship, or the need to remove it
- EntityIdInterface is the common interface, so when it's used it means the relationship is optional

e introduced this concept a bit late during our development but it appeared to be quite useful when we starting using it which is why we try to generalize it especially on use cases where they are useful.

Reply:
Oh and another advantage of this representation is that for now it relies on the 0 value in DB, but it's not ideal Using null in the DB would be cleaner So if one day we decide to change the DB structure we can simply update the const value and all the logic will remain the same

Reply:
Ok @jolelievre ! I got it, thanks a lot for explanations (:heart:) I think I understood what disturbed me in this architecture ^^

Let me know if I didn't understand.

- Development starts from code base with traditionnal relationnal description of entities.
- A CQRS system has been installed to manage migration.
- So we need some dedicated classes to describe and manage actions (query/commands).
- But we want to keep it the more generic we can (especially considering state/data models).

I think the problem I have with this architecture is about the confusion there is between the action and the state. And in fact you say it yourself :
> As such we need a value to identify the no relation, we can't rely on null because most of the setter in product commands use null to indicate that there is no updates. So we need those three use case:
>
>     * null: no modifications
>     * EntityId: set the relationship with specified entity
>     * NoEntityId: remove the relationship to any entity

Then I think what you are looking for is a strategy with parameters.

What do you think about this ?


```php

interface EditLinkInterface
{
    // ...
}

class UpdateLink implements EditLinkInterface
{
    // ...
}

class KeepLink implements EditLinkInterface
{
    // ...
}

class RemoveLink implements EditLinkInterface
{
    // ...
}

// ======================================= //
// ...
// ======================================= //

class EditableLinkStrategyInterface
{
    const KEEP_DEFAULT_VALUE = null; 
    const REMOVE_DEFAULT_VALUE = 0; 
}

class EditableLinkStrategy implements EditableLinkStrategyInterface
{
    public static function editLinkAction(?int $value): EditLinkInterface {
        switch($value) {
            case EditableLinkStrategyInterface::KEEP_DEFAULT_VALUE:
                return new KeepLink(...);
            case EditableLinkStrategyInterface::REMOVE_DEFAULT_VALUE:
                return new RemoveLink(...);
            default:
                return new UpdateLink($value);
        }
    }
}
```

With this, you can replace in any code this way :
```php
private AddressId $addressId;
private CustomerId $customerId;
private CountryId $countryId;
private StateId $stateId;

public function __construct(
    AddressId $addressId,
    CustomerId $customerId,
    CountryId $countryId,
    StateId $stateId,
) {
    $this->addressId = $addressId;
    $this->customerId = $customerId;
    $this->countryId = $countryId;
    $this->stateId = $stateId;
}
```


```php
private EditLinkInterface $addressId;
private EditLinkInterface $customerId;
private EditLinkInterface $countryId;
private EditLinkInterface $stateId;

public function __construct(
    AddressId $addressId,
    CustomerId $customerId,
    CountryId $countryId,
    StateId $stateId,
) {
    $this->addressLink = EditableLinkStrategy::editLinkAction($addressId->getValue());
    $this->customerLink = EditableLinkStrategy::editLinkAction($customerId->getValue());
    $this->countryLink = EditableLinkStrategy::editLinkAction($countryId->getValue());
    $this->stateLink = EditableLinkStrategy::editLinkAction($stateId->getValue());
}
```

And then when you want to give some meaning to your class for "routing" / building a strategy, you may use this : https://stackoverflow.com/questions/37029231/switch-by-class-instanceof-in-php

And if the generic Strategy doesn't work for you, you even can implement a specific Strategy for your entity.

I think I am still far from understanding the whole problem you are trying to solve but let me know what you think about the direction I am trying to show.

> :point_up: with this proposal we lost some type checking about link "nature" but I think we could balance [php lack of generics](https://wiki.php.net/rfc/generics#proposed_php_version_s) with some custom reflexion usage.
> 
> For exemple adding some stuff like this :
> 
> ```php
> interface EditLinkInterface
> {
>   public function getTargetEntityType(): string;
>   public function getTargetEntity(): mixed;
> }
> ```
> And adding proper asbtract behavior to all objects. 
> For exemple final usage may look `EditableLinkStrategy::editLinkAction($stateId->getValue(), $stateId::class)` or `EditableLinkStrategy::editLinkAction($stateId->getValue(), $stateId)`, ...

Reply:
I get the idea, and it could be interesting if we had multiple use cases but it seems a bit overkill for our need The real need behind this refacto is mainly: VOs shouldn't accept 0 value

For specific use case where we still need to use the 0 value we use a NoVOId It's simple and efficient enough IMHO

Reply:
What about this debate ? Can you continue, and go on with the path we initially choose ? or switch to the way that @MeKeyCool propose. Thx

Reply:
We discussed with @matks who explained me that the need is to be able to use `0` values in different situation for different purposes (so naming by `edit`, `keep`, ... won't be satisfying).

I am still not convinced that paying back this debt requires something that heavy to implement but from what I understood, @jolelievre and @matks think my proposal is too complicated ^^' (maybe I didn't explain it properly ... ^^')

I am open for discussion, If i receive some clear questions, I would enjoy to answer. But if it blocks everything because of lack of time, I won't insist more ^^'